File: Balanced_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int heightTree(TreeNode* root) {
        if( root == NULL) return 0;
        int lHeight = heightTree(root->left);
        if(lHeight == -1) return -1;
        int rHeight = heightTree(root->right);
        if(rHeight == -1) return -1;
        if(abs(lHeight-rHeight) > 1) return -1;
        return max(lHeight, rHeight)+1;
    }
    
    bool isBalanced(TreeNode* root) {
        return heightTree(root) != -1;
    }
};

File: balanced_binary_tree.java

class Solution {
    Boolean balanced = true;

    public int dfs(TreeNode root) {
        if(!balanced) return 0;

        if(root == null) return 0;
        int leftHeight = dfs(root.left);
        int rightHeight = dfs(root.right);
        // System.out.println(root.val + " " + Math.abs(leftHeight - rightHeight));
        if(Math.abs(leftHeight - rightHeight) > 1) {
            balanced = false;
            return 0;
        }
        return 1 + Math.max(leftHeight, rightHeight);
    }

    public boolean isBalanced(TreeNode root) {
        dfs(root);
        return balanced;
    }
}